### ISSUE TITLE
check.map crash

### ISSUE THREAD

**Issue by graetzer** [philbooth/check-types.js#41]
I think the [mapComplex](https://github.com/philbooth/check-types.js/blob/master/src/check-types.js#L490) function will cause the ```check.map``` call to crash, in case one specifies something like ```check.map({}, { key : { subkey : check.string }})```

Not sure if it's intended this way, but shouldn't the function check if ```data``` is defined before calling ```data[key]``` ?

    function mapComplex (data, predicates) {
        var result = {};
        Object.keys(predicates).forEach(function (key) {
            var predicate = predicates[key];
            if (isFunction(predicate)) {
                result[key] = predicate(data[key]);
            } else if (object(predicate)) {
                result[key] = mapComplex(data[key], predicate);
            }
         });
        return result;
    }

**Comment by philbooth** [comment]
Yep, you're right. It's definitely a bug and it's definitely a gap in the test coverage. Thanks for pointing it out.

There is a bit of a back-log of stuff building up on this project now I realise, so I'll make some time to tackle this and some of the other issues soon, maybe Sunday. Or of course if you've got time to submit a pull request yourself at any point, that is always appreciated too. :)

**Comment by graetzer** [comment]
I applied a quick fix to my code. In case the parameter```data``` is undefined I'm applying the predicate to data instead.
My Reasoning here is that I still want the behaviour of the ```maybe``` modifier for my subkeys.
   
     function mapComplex (data, predicates) {
        var result = {};
        Object.keys(predicates).forEach(function (key) {
            var predicate = predicates[key];
            if (isFunction(predicate)) {
                if (!data) {
                    result[key] = predicate(data);
                } else {
                    result[key] = predicate(data[key]);
                }
            } else if (object(predicate)) {
                result[key] = mapComplex(data ? data[key] : data, predicate);
            }
         });
        return result;
    }

**Comment by philbooth** [comment]
Hey @graetzer, sorry for taking a while to get back to you. I didn't want to say anything until I'd had a chance to think about it properly.

I think there is one problem with your solution above. It makes sense for the `maybe` case that you mention, but the behaviour is too surprising if the predicate is a straightforward test for `undefined`:

```js
// Returns { foo: { bar: true } }
var result = check.map({}, { foo: { bar: check.undefined } });

// I would expect the following assertion to hold, otherwise
// the predicate would have been applied to `foo` directly.
assert.isFalse(result.foo.bar);
```

So, if we detect `check.undefined` or `check.not.assigned`, I think we should negate the result from your modification. Otherwise I think the result of your modification should be returned as is.

What do you reckon?

**Comment by philbooth** [comment]
Fwiw, I've pushed a branch containing two unit tests that cover this. If you're up for opening a PR, these are what need to pass:

https://github.com/philbooth/check-types.js/compare/issue_41?expand=1

:)

**Comment by philbooth** [comment]
Hey @graetzer, I've published a release containing this fix, version 3.2.0. I hope it works how you want it to, but if not let me know and I'll sort it out! :)
